src/swarm/neo/connection/RequestOnConnBase.d

```diff
@@ -507,6 +614,185 @@ abstract class RequestOnConnBase
+                    If this argument is null, sending does not occur
+
+            Returns:
+                an EventNotification instance denoting the event which occurred
+
+            Throws:
+                Exception on protocol or I/O error.
+
+        ***********************************************************************/
+
+        public EventNotification nextEvent (
+            Flags flags, FillPayloadDg fill_payload = null )
+        out ( fired_event )
+        {
+            auto fired_event_non_const = cast(EventNotification)fired_event;
+            assert(fired_event_non_const.active !=
+                fired_event_non_const.active.none);
+        }
+        body
+        {
+            EventNotification fired_event;
+
+            bool sending = fill_payload !is null;
+            bool receiving = (flags & Flags.Receive) > 0;
+            bool yielding = (flags & Flags.Yield) > 0;
+
+            // Set up sending, if required.
+            scope payload = this.new Payload;
+
+            if ( sending )
+            {
+                fill_payload(payload);
+                this.outer.send_payload_ = this.outer.send_payload;
+
+                // registerForSending may fail if called while connection
+                // shutdown is in progress
+                scope ( failure )
+                    this.outer.send_payload_ = null;
+
+                if (!this.outer.connection.registerForSending(
+                    this.outer.request_id))
+                {
+                    assert(false, "nextEvent: already sending");
+                }
+
+                // Sending may succeed immediately. send_payload_ is null, in
+                // this case.
+                if ( !this.outer.send_payload_ )
+                {
+                    fired_event.sent = Sent();
+                    return fired_event;
+                }
+            }
+
+            scope ( exit )
+            {
+                if ( sending )
+                    this.outer.connection.unregisterForSending(
+                        this.outer.request_id);
+            }
+
+            // Set up receiving, if required.
+            if ( receiving )
+            {
+                // registerForErrorNotification may fail if called while
+                // connection shutdown is in progress
+                scope ( failure )
+                    this.outer.recv_payload_ = null;
+
+                if (!this.outer.connection.registerForErrorNotification(
+                    this.outer.request_id))
+                {
+                    assert(false, "nextEvent: already receiving");
+                }
+            }
+
+            scope ( exit )
+            {
+                if ( receiving )
+                    this.outer.connection.unregisterForErrorNotification(
+                        this.outer.request_id);
+            }
+
+            // Set up yielding, if required.
+            scope yielded_rqonconn = this.outer.new YieldedRequestOnConn;
```

Comment:
Too bad there's no nice way to escape creating `yielded_rqonconn` and `payload` even if you don't need them.